### ISSUE TITLE
player.seekTo vs presentationTimeUs in overlay.getTextureId is wrong

### ISSUE THREAD

**Issue by steelbytes** [androidx/media#1996]
### Version

Media3 1.5.1

### Devices that reproduce the issue

pixel 8pro

### Reproducible in the demo app?

Not tested

### Reproduction steps

```
player.playWhenReady = false
player.setVideoEffects(myEffects)
player.setMediaItem(mediaItem)
player.prepare()
player.seekTo(exoPos)
watch presentationTimeUs in myTextureOverlay.getTextureId()
```

### Expected result

presentationTimeUs in textureOverlay.getTextureId should start counting from the start of the file. 

### Actual result

presentationTimeUs in textureOverlay.getTextureId should from the seek position.

### Media

all files tested

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to android-media-github@google.com after filing this issue.

**Comment by steelbytes** [comment]
can't test in demo app as can't compile your git repo on windows with AS koala.  I used to be able to with older pulls.

**Comment by claincly** [comment]
> Expected result:  presentationTimeUs in textureOverlay.getTextureId should start counting from the start of the file.

This is not the expected behaviour. The effect timestamp will start from `exoPos` instead, if you want the timestamp to start from zero, you need to manually subtract the offset in your effect.

**Comment by steelbytes** [comment]
well it reall

> > Expected result:  presentationTimeUs in textureOverlay.getTextureId should start counting from the start of the file.
> 
> This is not the expected behaviour. The effect timestamp will start from `exoPos` instead, if you want the timestamp to start from zero, you need to manually subtract the offset in your effect.

:-(
it feels like a *bug*.  I haven't checked but is that the way it's documented?

**Comment by claincly** [comment]
No it's not formally documented, but the implementation follows the principle of ExoPlayer, where after you seek, the position is reset to the seeked position.

**Comment by steelbytes** [comment]
well trying to workaround this produces inconsistent results.  guess I'm just going to have to fork it.

**Comment by steelbytes** [comment]
AND presentationTimeUs is sometimes negative.

**Comment by claincly** [comment]
The `presentationTimeUs` should not be negative - could you share how did you trigger this to happen?

**Comment by steelbytes** [comment]
> The `presentationTimeUs` should not be negative - could you share how did you trigger this to happen?

I have a hunch that it happens when I to the above use of seekTo AND then bash the skip backwards/forwards and seek bar a bit.

simply throwing an assert if (presentationTimeUs < 0L) does trigger.

**Comment by steelbytes** [comment]
ok, here's a trivial repro for negative time

1. pull https://github.com/steelbytes/media3test
2. add overlay code below to the existing list (can remove Crop)
3. add player.seekTo(60*1000L) after prepare
4. adb logcat -e negative -T 0
5. compile & run
6. click the go to start arrow in the ui (standard controller)
7. wait for it to load and render the frame
8. observe logcat and yo'ull see the negative

```
                OverlayEffect(ImmutableList.Builder<TextureOverlay>().add(object : TextOverlay() {
                    override fun getText(presentationTimeUs: Long): SpannableString = SpannableString("test")
                    override fun getTextureId(presentationTimeUs: Long): Int {
                        if (presentationTimeUs < 0L)
                            Log.d("Media3Test", "negative presentationTimeUs " + presentationTimeUs)
                        return super.getTextureId(presentationTimeUs)
                    }
                }).build())

```

**Comment by steelbytes** [comment]
Bump for the negative time

**Comment by kim-vde** [comment]
Coming back to the original issue (I want to make sure I understand first before tackling the negative timestamps problem), if you have a clip of 10 seconds and seek to 5 seconds, what is the first timestamp you would expect `getTextureId()` to receive after the seek? The current behavior is that the first timestamp is 5 seconds, which is what makes the most sense to me.

**Comment by steelbytes** [comment]
>  if you have a clip of 10 seconds and seek to 5 seconds, what is the first timestamp you would expect `getTextureId()` to receive after the seek? The current behavior is that the first timestamp is 5 seconds

it isn't.

the first timestamp in getTextureId is 0 if I seekTo.

```
                player.setVideoEffects(
                    listOf<Effect>(
                        OverlayEffect(ImmutableList.Builder<TextureOverlay>().add(object : TextOverlay() {
                            override fun getText(presentationTimeUs: Long): SpannableString = SpannableString("test")
                            override fun getTextureId(presentationTimeUs: Long): Int {
                                Log.d("Media3Test", "presentationTimeUsTest overlay " + (presentationTimeUs/1000.0).roundToLong() + "ms")
                                return super.getTextureId(presentationTimeUs)
                            }
                        }).build())
                    )
                )
                Log.d("Media3Test", "presentationTimeUsTest media " + uri.path)
                player.setMediaItem(MediaItem.Builder().setUri(uri).build())
                player.prepare()
                Log.d("Media3Test", "presentationTimeUsTest seek 60000ms")
                player.seekTo(60000L)
                lifecycleScope.launch {
                    delay(1000)
                    Log.d("Media3Test", "presentationTimeUsTest seek 61000ms")
                    player.seekTo(61000L)
                    lifecycleScope.launch {
                        delay(1000)
                        Log.d("Media3Test", "presentationTimeUsTest seek 59000ms")
                        player.seekTo(59000L)
                    }
                }
```

```
10:31:24.556 24611-24611 Media3Test           com...tes.media3test presentationTimeUsTest media /document/primary:test/VID-20230803131443-GX010174.MP4
10:31:24.560 24611-24611 Media3Test           com...tes.media3test presentationTimeUsTest seek 60000ms
10:31:24.933 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay 0ms
10:31:24.951 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay 40ms
10:31:25.565 24611-24611 Media3Test           com...tes.media3test presentationTimeUsTest seek 61000ms
10:31:25.945 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay 1000ms
10:31:25.977 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay 1040ms
10:31:26.575 24611-24611 Media3Test           com...tes.media3test presentationTimeUsTest seek 59000ms
10:31:27.070 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay -1000ms
10:31:28.093 24611-24745 Media3Test           com...tes.media3test presentationTimeUsTest overlay -960ms
```

**Comment by claincly** [comment]
My apologies - I read your issue completely the other way around, I thought you mean seeing zero after seeking to 5 is the expected behaviour, and I have assumed you are using `CompositionPlayer`.

Yes I can confirm with your code, the `Overlay` does receive timestamp 0 after the seek. This is a known limitation of ExoPlayer's timeline though. The problem is if you seek before the player becomes `READY`, the player would assume the starting point of the media is the seek point. More concretely, if you have a clip 10s long, and if you seek to 5s before the player is `READY`, the player would assume the media starts at 5s and treat 5s as the starting position, and this way your effect would receive 0 after seek. Likewise, if you seek to zero now, the timestamp will be -5.

There's no fix to this issue unless ExoPlayer changes the way the timeline is handled. What you could try though, are

- Try seeking after the player is `READY`, you could do it in the player listener, like 
```
    @Override
    public void onPlaybackStateChanged(@Player.State int playbackState) {
      if (playbackState == Player.STATE_READY) {
          // Make sure this is called only once.
          player.seekTo(60_000);
        }
    }

```
-  Or switch to using `CompositionPlayer`

Note with ExoPlayer, if you have more than one `MediaItem` in the playlist, the seek timestamps will look weird if seeked across MediaItem boundaries. The best approach is to try `CompositionPlayer`.

`CompositionPlayer` is under active development and can be unstable.

**Comment by steelbytes** [comment]
> * Or switch to using `CompositionPlayer`

where are the docs? no mention of this Player in either https://developer.android.com/media/media3 or https://developer.android.com/reference/androidx/media3/common/package-summary

It has `@RestrictTo(LIBRARY_GROUP)` so can't use it anyway?

PS why is this in transformer package? seems weird place.

**Comment by kim-vde** [comment]
`CompositionPlayer` is still work-in-progress. It hasn't been officially released yet but it's already available for people who want to try it. The annotation advise you to not use it (because it hasn't been officially released), but you can still use it if you want to.

**Comment by steelbytes** [comment]
CompositionPlayer does seem to fix the issues with timestamps.  It also fixes the memory leak I reported separately. 

Annoyingly I have to extract the duration of the video using something like MediaMetadataRetriever and then call EditedMediaItem.Builder.setDurationUs.  Messy, feels fragile and dodgy.

Also the goto start/end/next buttons on the controller are disabled, the goto start works with exoplayer (I only am loading one file).  The -5/+15 work fine.

**Comment by kim-vde** [comment]
> Annoyingly I have to extract the duration of the video using something like MediaMetadataRetriever and then call EditedMediaItem.Builder.setDurationUs.

This is something that we intend to improve in the future. We don't have the capacity to do it soon though.

> Also the goto start/end/next buttons on the controller are disabled, the goto start works with exoplayer (I only am loading one file).

Could you please open a separate issue for this, with some reproduction steps. Thanks.
